[text-wrap-style pretty] Refactor cost function and entry data struct.

yulunwu2 opened the pull request:
#### 79a60c46a38e1755de8489acaa6349acad80f9cb
<pre>
[text-wrap-style pretty] Refactor cost function and entry data struct.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287177">https://bugs.webkit.org/show_bug.cgi?id=287177</a>
&lt;<a href="https://rdar.apple.com/144333739">rdar://144333739</a>&gt;

Reviewed by Alan Baradlay.

This is a cleanup/refactor PR that does a couple things:

1) Refactors out the float adjusting text-wrap-style:pretty target line length.
This makes it easier to find and adjust this float in the future.

2) Refactors out `entry` struct into `entryBalance` and `entryPretty` for clarity.
This removes duplicate code shared between balanceRangeWithLineRequirement and
balanceRangeWithoutLineRequirement, as well as clarifies that prettifyRange a
different struct. Note that entryPretty:endsWithHyphen is currently not used. This
will be fixed in a follow up PR.

3) Refactors out cost computation for pretty, balance, and raggedness.
This clarifies the scoring use for text-wrap-style:balance and text-wrap-style:pretty,
making it easier for future authors to understand what heuristics are being used
for different text-wrap-styles and how the differ. Additionally, this also clarifies
the use of the raggedness metric which is used by both text-wrap-mode:balance as well
as the first line for text-wrap-mode:pretty.

Finally, this PR adds FIXMEs for computeCostPretty, clarifying opportunities for future work.

Canonical link: <a href="https://commits.webkit.org/290053@main">https://commits.webkit.org/290053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e52e24e57f376e8565a4b29965b53a20afa78ae4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43032 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93547 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39341 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90638 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16290 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68306 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26016 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6496 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80121 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48670 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6263 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34535 "Found 3 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38449 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76633 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35425 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95387 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15762 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11540 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77172 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75978 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76441 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18882 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20831 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19259 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/8801 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15778 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21086 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17301 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -42,12 +42,51 @@ namespace Layout {
+// overshoot and undershoot the target line length, giving more flexibility in the solution search.
+static const float textWrapPrettyMaximumLineWidthAdjustment = 0.95;
+
+struct entryBalance {
```
this should be named EntryBalance (and EntryPretty)  (surprised that style checker didn't complain)

yulunwu2 replied:
Done, thanks!

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
```diff
@@ -42,12 +42,51 @@ namespace Layout {
+static const float textWrapPrettyMaximumLineWidthAdjustment = 0.95;
+
+struct entryBalance {
+    float accumulatedCost { std::numeric_limits<float>::infinity() };
```
I know this just moves some code around but it very much looks like a magic number.

yulunwu2 replied:
I assume this is RE: 

static const float textWrapPrettyMaximumLineWidthAdjustment = 0.95;

I actually really don't like this magic number.

The ideal line width used for balance:
            
https://github.com/WebKit/WebKit/blob/96ace0a27658cb30e164623e4915ab2e0335441f/Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp#L221

looks far better than:

https://github.com/WebKit/WebKit/blob/96ace0a27658cb30e164623e4915ab2e0335441f/Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp#L230

so I plan on removing it in a follow up CL. 

However, I *don't* think we'll be able to get away from magic numbers. I plan on rewriting computeRaggedness as per:

http://www.eprg.org/G53DOC/pdfs/knuth-plass-breaking.pdf

In particular, the cost function:

        Thejth line also has a total stretchability Yj and
        total shrinkability Zj, obtained by summing all of the yi and zi for glue items in the
        range uj < i < bj. Now we can compare the actual length Lj to the desired length Z,
        by seeing if there is enough stretchability or shrinkability to change Lj into 4; we
        define the adjustment ratio rj of the jth line as follows:
        If Lj = lj (a perfect fit), let rj = 0.
        If L, < Z j (a short line), let r, = (Zj-Lj)/ Yj, assuming that Y j > 0; the value
        of rj is undefined if Yj< 0 in this case.
        If L, > Z, (a long line), let r j = (Zj-Lj)/Zj, assuming that Z j > 0; the value of r j
        is undefined if Z j < 0 in this case.
        
Allows for different penalties for stretching and shrinking relative to the ideal line width. These would all be magic numbers (as is the way penalties for consecutive hyphenations are done, it is yet another random number).

yulunwu2 replied:
BTW happy to have a discussion offline about the use of...magical numbers

yulunwu2 committed:
[text-wrap-style pretty] Refactor cost function and entry data struct.
https://bugs.webkit.org/show_bug.cgi?id=287177
<rdar://144333739>

Reviewed by Alan Baradlay.

This is a cleanup/refactor PR that does a couple things:

1) Refactors out the float adjusting text-wrap-style:pretty target line length.
This makes it easier to find and adjust this float in the future.

2) Refactors out `entry` struct into `entryBalance` and `entryPretty` for clarity.
This removes duplicate code shared between balanceRangeWithLineRequirement and
balanceRangeWithoutLineRequirement, as well as clarifies that prettifyRange a
different struct. Note that entryPretty:endsWithHyphen is currently not used. This
will be fixed in a follow up PR.

3) Refactors out cost computation for pretty, balance, and raggedness.
This clarifies the scoring use for text-wrap-style:balance and text-wrap-style:pretty,
making it easier for future authors to understand what heuristics are being used
for different text-wrap-styles and how the differ. Additionally, this also clarifies
the use of the raggedness metric which is used by both text-wrap-mode:balance as well
as the first line for text-wrap-mode:pretty.

Finally, this PR adds FIXMEs for computeCostPretty, clarifying opportunities for future work.

Canonical link: https://commits.webkit.org/290053@main
